crypto/evp/p_lib.c

```diff
@@ -317,78 +318,234 @@ int EVP_PKEY_cmp(const EVP_PKEY *a, const EVP_PKEY *b)
 }
 
-EVP_PKEY *EVP_PKEY_new_raw_private_key(int type, ENGINE *e,
-                                       const unsigned char *priv,
-                                       size_t len)
+
+static EVP_PKEY *new_raw_key_int(OPENSSL_CTX *libctx,
+                                 const char *strtype,
+                                 const char *propq,
+                                 int nidtype,
+                                 ENGINE *e,
+                                 const unsigned char *key,
+                                 size_t len,
+                                 int priv)
 {
-    EVP_PKEY *ret = EVP_PKEY_new();
+    EVP_PKEY *pkey = NULL;
+    EVP_PKEY_CTX *ctx = NULL;
+    OSSL_PARAM_BLD *tmpl = NULL;
+    OSSL_PARAM *params = NULL;
+    const EVP_PKEY_ASN1_METHOD *ameth = NULL;
+    int result = 0;
 
-    if (ret == NULL
-        || !pkey_set_type(ret, e, type, NULL, -1, NULL)) {
-        /* EVPerr already called */
+    /* Check if there is an Engine for this type */
+    if (e == NULL) {
+        ENGINE *tmpe = NULL;
+
+        if (strtype != NULL)
+            ameth = EVP_PKEY_asn1_find_str(&tmpe, strtype, -1);
+        else if (nidtype != EVP_PKEY_NONE)
+            ameth = EVP_PKEY_asn1_find(&tmpe, nidtype);
+
+        /* If tmpe is NULL then no engine is claiming to support this type */
+        if (tmpe == NULL)
+            ameth = NULL;
+
+        ENGINE_finish(tmpe);
+    }
+
+    if (e == NULL && ameth == NULL) {
+        /*
+         * No engine is claiming to support this type, so lets see if we have
+         * a provider.
+         */
+        ctx = EVP_PKEY_CTX_new_from_name(libctx,
+                                         strtype != NULL ? strtype
+                                                         : OBJ_nid2sn(nidtype),
+                                         propq);
+        if (ctx == NULL) {
+            EVPerr(0, ERR_R_MALLOC_FAILURE);
+            goto err;
+        }
+        if (EVP_PKEY_key_fromdata_init(ctx) == 1) {
+            tmpl = OSSL_PARAM_BLD_new();
+            if (tmpl == NULL) {
+                EVPerr(0, ERR_R_MALLOC_FAILURE);
+                goto err;
+            }
+
+            if (!priv && !OSSL_PARAM_BLD_push_octet_string(tmpl,
+                                                           OSSL_PKEY_PARAM_PUB_KEY,
+                                                           key, len)) {
+                EVPerr(0, EVP_R_KEY_SETUP_FAILED);
+                goto err;
+            }
+
+            if (priv && !OSSL_PARAM_BLD_push_octet_string(tmpl,
+                                                          OSSL_PKEY_PARAM_PRIV_KEY,
+                                                          key, len)) {
+                EVPerr(0, EVP_R_KEY_SETUP_FAILED);
+                goto err;
+            }
+
+            params = OSSL_PARAM_BLD_to_param(tmpl);
+            if (params == NULL) {
+                EVPerr(0, ERR_R_MALLOC_FAILURE);
+                goto err;
+            }
```

Comment:
Er, I fail to see why you go through all the trouble with OSSL_PARAM_BUILD here...

``` C
        OSSL_PARAM params[] = { OSSL_PARAM_END, OSSL_PARAM_END };

        params[0] = OSSL_PARAM_construct_octet_string(priv ? OSSL_PKEY_PARAM_PRIV_KEY
                                                           : OSSL_PKEY_PARAM_PUB_KEY,
                                                      key, len);
```

Done.

Reply:
Yes - that's much better. Done.